rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,103 @@
+
+| action       | initial state | change                           |
+| ------------ | ------------- | -------------------------------- |
+| role removal | _admin_       | _admin_ tries to remove own role |
+
+The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by manually instrumenting the auth process using tctl.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
+2. User with a role that grants user/role editing capabilities is _admin_.
```

Comment:
> Please correct me if I am wrong, isn't role editing capabilities (role with create/update rules to the role resource) just enough?

If there is only one user that can edit/create/delete new roles (no other user), he cannot create a new user with a role or assign a role to some other user. He is locked. So the _admin_ should have a combination of user/role editing capabilities

Reply:
> If an admin lost user editing capabilities while retaining role editing capabilities, they can still edit the role and add back those user editing capabilities
Right, because he still has at least one role which he can edit 👍🏻